### DISCUSSION TITLE
Some Elementary Questions

### DISCUSSION THREAD

**Discussion by 2024AfD** [rems-project/sail#846] [Q&A]
Consider the following invocation, which comes from the Makefile of the Sail model for RISC-V.

     sail ... -c_preserve _set_Misa_C ... -o generated_definitions/c/riscv_model_RV64.c

where "..." represents the other flags.

sail --help explains

     --c-preserve

without mentioning -c_preserve.

Question 1:  In general, if --xx-yy is a flag, then can we use it by specifying -xx_yy ?

Below is the explanation for --c-preserve.

     make sure the provided function identifier is preserved in C output

Below are the only lines (in riscv_model_RV64.c) that contain _set_Misa_C.

     unit z_set_Misa_C(struct zMisa*, uint64_t);
     void startup_z_set_Misa_C(void)
     unit z_set_Misa_C(struct zMisa* zr_ref, uint64_t zv)
     void finish_z_set_Misa_C(void)
       startup_z_set_Misa_C();
       finish_z_set_Misa_C();

Question 2:  Should the explanation for --c-preserve be interpreted to mean "make sure all function identifiers with the provided suffix is preserved in C output"?

**Comment by Alasdair** [comment]
> Question 1: In general, if --xx-yy is a flag, then can we use it by specifying -xx_yy ?

Yes, the argument parser is flexible in what it accepts - but this is mostly for backwards compatibility. I would suggest using the more standard `--long-option` form.

> Question 2: Should the explanation for --c-preserve be interpreted to mean "make sure all function identifiers with the provided suffix is preserved in C output"?

`--c-preserve <identifier>` means the _Sail_ identifier (and everything it depends on) will be preserved into the C output, it may still be name-mangled, and it might bring auxiliary definitions with it like the `startup_` and `finish_` functions. When we compile to C we do dead code elimination of everything that isn't reachable from `main` (and some other root functions like `decode` and `execute`). What the `--c-preserve` flag does internally is it adds the given identifier to the set of roots for that dead-code elimination pass.
